### ISSUE TITLE
Additional languages

### ISSUE THREAD

**Issue by MisterY** [dslul/openboard#31]
Some languages (like Bosnian) are not on the list. How does it work with adding them?
Gboard has tons and I'm not sure about AOSP.
As a workaround, I could use any other language definition, since the dictionaries seem to be empty. I.e. #18 

Related: #24

**Comment by MisterY** [comment]
Btw, thanks for the instructions on how to create a dictionary! Now it's just about adding the language(s) itself.

**Comment by foul-fortune-feline** [comment]
日本語をください！
Japanese please!

**Comment by USB3-2-10Gbit** [comment]
i think more languages should be added by default.

I have read from some users who want to have dictionaries for vietnamese, japanese etc...

And a search button is missing.

Another question how big would the app be if all languages were added? In general, smartphones have more and more internal storage in the last few months.

Maybe an external function can be added, namely access to storage. Users can simply load the dictionaries and add them themselves using standard file manager.

**Comment by MisterY** [comment]
Agree, it would be perfect if only selected languages could be installed / downloaded / copied.
Also, synchronizing them over multiple devices, even if manually, would be great and would save tons of time training the other instances.

**Comment by jwsp1** [comment]
> Agree, it would be perfect if only selected languages could be installed / downloaded / copied.
> Also, synchronizing them over multiple devices, even if manually, would be great and would save tons of time training the other instances.

AFAIK downloading packages afterwards is not possible since the app has no network permissions (which is the way it should be, due to privacy concerns!). 
A solution could be providing a second app which downloads dictionaries/languages. The main app will recognize the downloaded files.

**Comment by MisterY** [comment]
> AFAIK downloading packages afterwards is not possible since the app has no network permissions (which is the way it should be, due to privacy concerns!).
> A solution could be providing a second app which downloads dictionaries/languages. The main app will recognize the downloaded files.

I'm not saying an app should do it but that such an ability would be fantastic. I don't mind downloading and copying the files myself if there are links provided, for example.

**Comment by MisterY** [comment]
Is it too complicated to publish dictionaries separately? Each file is 1MB or larger. Not downloading the dictionaries which will never be used would slim the app size significantly.

**Comment by USB3-2-10Gbit** [comment]
> Is it too complicated to publish dictionaries separately? Each file is 1MB or larger. Not downloading the dictionaries which will never be used would slim the app size significantly.

I don't like it when the app gets smaller.

Gboard for example has extremely many dictionaries integrated. The advantage is that you don't have to download again and can use everything locally and offline. The apk file is bigger but very tiny compared to the storage of current Android devices.

SwiftKey, for example, always has to reload the dictionaries and because OpenBoard is a privacy app. Should not give access to the internet.

In general there is already 5G or 4G connection and that with unlimited data plans. There are more and more Android smartphones with USB 3 port to back up the data or apk file from the device to external hard disk or to the PC.

Also the integrated storage for Android devices is getting bigger and bigger and it doesn't cost much.

**Comment by dslul** [comment]
I'm still not sure how to handle this. On one hand, regular users don't want to deal with additional apps to download missing bits, on the other people that care about privacy don't want internet access. Right now the apk is ~30 MiB, not much in my opinion and it can still grow a bit, but if we add a lot more dictionaries it can become unfeasible to embed them all.

**Comment by MisterY** [comment]
Yes, that's a very interesting problem. I hope that the granular control of the permissions in the current Android versions can help solve this. The users can control whether they grant or revoke the network permission and the app can request it only when downloading the dictionaries.
This would be quite acceptable for me, as I could revoke the network permission once the desired dictionaries are downloaded. 
That could be one option.

**Comment by jwsp1** [comment]
I would go with the solution of providing 10 majorly requested languages and having another app which can handle the download/deletion of languages packages if necessary

**Comment by MisterY** [comment]
I think this is why some other apps offer per-dictionary download. This is also a decent option. One should be able to download an apk or install from their favourite app store the languages they need.
Packaging separate dictionaries as individual apps/libraries would solve this.

**Comment by rancidfrog** [comment]
The default app that people should use is the browser!
Which already has internet permission.
App should be able to import dictionaries.
Internet not required as dictionaries downloaded by browser/system downloader.
App should be able to manage ductionaries, import/enable/disable/delete/etc
Dictionary packs can be _released_ in github.

**Comment by flinstones1** [comment]
@dslul Hello, I'm a very novice programmer. I'd like to try and implement Korean. I had a look at how other languages are implemented and couldn't find specific logic for how character formation. (By logic I mean: In some languages, a character is updated when another character is typed next to it. Eg. In Thai some characters can stack on top of other letters)
The only language specific code I found were implementing of the unicode blocks. Is this all that is necessary or am I missing something?

**Comment by Lee0701** [comment]
@flinstones1 Hi, I've been implementing Korean input logic and layouts on my fork, so please check if you're interested. #24
